File: web/src/lib/converters/pvr-to-png.ts
import {
  ConverterInfo,
  ConversionOptions,
  ConversionResult,
  IConverter,
  defaultValidate,
  DesktopRequiredError,
} from "./interfaces";

// PVR (PowerVR Texture) v3 decoder.
// (技术规范 §4.3.1 names pvr-texture-decoder-wasm; here we ship a self-contained
// decoder that handles uncompressed 32-bit RGBA PVR v3 files with no external
// dependency, and degrades gracefully for compressed variants.)
const PVR3_MAGIC = 0x03525650;

interface PvrHeader {
  version: number;
  flags: number;
  pixelFormat: bigint;
  colorSpace: number;
  channelType: number;
  height: number;
  width: number;
  depth: number;
  numSurfaces: number;
  numFaces: number;
  numMipmaps: number;
  metaDataSize: number;
}

function parsePvrHeader(buffer: ArrayBuffer): PvrHeader {
  const dv = new DataView(buffer);
  return {
    version: dv.getUint32(0, true),
    flags: dv.getUint32(4, true),
    pixelFormat: dv.getBigUint64(8, true),
    colorSpace: dv.getUint32(16, true),
    channelType: dv.getUint32(20, true),
    height: dv.getUint32(24, true),
    width: dv.getUint32(28, true),
    depth: dv.getUint32(32, true),
    numSurfaces: dv.getUint32(36, true),
    numFaces: dv.getUint32(40, true),
    numMipmaps: dv.getUint32(44, true),
    metaDataSize: dv.getUint32(48, true),
  };
}

export class PvrToPngConverter implements IConverter {
  readonly info: ConverterInfo = {
    id: "pvr-to-png",
    name: "PVR to PNG",
    sourceFormats: [".pvr"],
    targetFormat: ".png",
    category: "image",
    maxWebFileSize: 100 * 1024 * 1024,
    classType: "A",
    description: "Convert PowerVR textures to PNG images",
  };

  validate(file: File) {
    return defaultValidate(this.info, file);
  }

  async convert(options: ConversionOptions): Promise<ConversionResult> {
    const buffer = await options.inputFile.arrayBuffer();
    const header = parsePvrHeader(buffer);

    if (header.version !== PVR3_MAGIC) {
      throw new DesktopRequiredError(
        "Not a valid PVR v3 file (bad magic number). Expected 'PVR\\3'.",
      );
    }

    const w = header.width;
    const h = header.height;
    if (w === 0 || h === 0) {
      throw new DesktopRequiredError("PVR header reports zero width/height.");
    }

    // First mipmap data starts at byte 52; metadata block is at the end.
    const dataStart = 52;
    const metaStart = buffer.byteLength - header.metaDataSize;
    const dataLen = metaStart - dataStart;
    const expected = w * h * 4;

    let rgba: Uint8ClampedArray;
    if (dataLen === expected) {
      // Heuristic: uncompressed 32-bit RGBA (R,G,B,A per pixel, little-endian).
      const view = new Uint8Array(buffer, dataStart, dataLen);
      rgba = new Uint8ClampedArray(view);
    } else {
      throw new DesktopRequiredError(
        `Compressed PVR variants (PVRTC/ETC/ASTC) are not decoded in-browser in this build. ` +
          `Data size (${dataLen} bytes) does not match uncompressed RGBA (${expected} bytes). ` +
          `Please use the free desktop app for compressed PVR files.`,
      );
    }

    const blob = await rgbaToPngBlob(rgba, w, h);
    return {
      data: blob,
      filename: options.inputFile.name.replace(/\.pvr$/i, ".png"),
      size: blob.size,
      mimeType: "image/png",
    };
  }
}

async function rgbaToPngBlob(
  rgba: Uint8ClampedArray,
  w: number,
  h: number,
): Promise<Blob> {
  const canvas = document.createElement("canvas");
  canvas.width = w;
  canvas.height = h;
  const ctx = canvas.getContext("2d");
  if (!ctx) throw new Error("Canvas 2D context unavailable");
  const img = ctx.createImageData(w, h);
  img.data.set(rgba);
  ctx.putImageData(img, 0, 0);
  return await new Promise<Blob>((resolve, reject) => {
    canvas.toBlob(
      (b) => (b ? resolve(b) : reject(new Error("PNG encoding failed"))),
      "image/png",
    );
  });
}

File: desktop/src-tauri/src/converters/pvr_to_png.rs
use std::fs;
use std::path::Path;

use image::{ImageFormat, RgbaImage};

use crate::converters::Converter;
use crate::errors::AppError;

/// Little-endian magic for PVR v3: bytes `50 56 52 03` = 'P' 'V' 'R' 0x03.
const PVR3_MAGIC: u32 = 0x0352_5650;
const HEADER_LEN: usize = 52;

/// PVR (PowerVR texture) -> PNG.
///
/// Implements a self-contained PVR v3 (52-byte header) parser. For the
/// uncompressed 32-bit RGBA layout we decode directly and re-encode as PNG.
/// Compressed PVRTC/ETC/ASTC variants are not decoded in this build and return a
/// clear `NotImplemented` so the UI can route the user to the planned native
/// decoder or suggest the web tool where applicable.
pub struct PvrToPngConverter;

impl Converter for PvrToPngConverter {
    fn slug(&self) -> &'static str {
        "pvr-to-png"
    }
    fn class_type(&self) -> &'static str {
        "A"
    }
    fn convert(&self, input: &Path, output: &Path) -> Result<u64, AppError> {
        let data = fs::read(input)?;
        if data.len() < HEADER_LEN + 16 {
            return Err(AppError::InvalidFile(
                "File too small to be a PVR v3 image.".into(),
            ));
        }

        let magic = u32::from_le_bytes([data[0], data[1], data[2], data[3]]);
        if magic != PVR3_MAGIC {
            return Err(AppError::InvalidFile(
                "Not a valid PVR v3 file (bad magic). Expected 'PVR\\3'.".into(),
            ));
        }

        // Header fields we need (all little-endian).
        let _flags = u32::from_le_bytes([data[4], data[5], data[6], data[7]]);
        let _pixel_format_lo = u32::from_le_bytes([data[8], data[9], data[10], data[11]]);
        let _pixel_format_hi = u32::from_le_bytes([data[12], data[13], data[14], data[15]]);
        let _color_space = u32::from_le_bytes([data[16], data[17], data[18], data[19]]);
        let _channel_type = u32::from_le_bytes([data[20], data[21], data[22], data[23]]);
        let height = u32::from_le_bytes([data[24], data[25], data[26], data[27]]);
        let width = u32::from_le_bytes([data[28], data[29], data[30], data[31]]);
        let _depth = u32::from_le_bytes([data[32], data[33], data[34], data[35]]);
        let _num_surfaces = u32::from_le_bytes([data[36], data[37], data[38], data[39]]);
        let _num_faces = u32::from_le_bytes([data[40], data[41], data[42], data[43]]);
        let _num_mipmaps = u32::from_le_bytes([data[44], data[45], data[46], data[47]]);
        let meta_data_size = u32::from_le_bytes([data[48], data[49], data[50], data[51]]);

        if width == 0 || height == 0 {
            return Err(AppError::InvalidFile(
                "PVR header reports zero width/height.".into(),
            ));
        }

        let data_start = HEADER_LEN;
        let meta_start = data.len().saturating_sub(meta_data_size as usize);
        let data_len = meta_start.saturating_sub(data_start);
        let expected = width as usize * height as usize * 4;

        if data_len != expected {
            return Err(AppError::NotImplemented(
                "Compressed PVR variants (PVRTC/ETC/ASTC) are not decoded in this build. \
                 Convert from an uncompressed 32-bit RGBA PVR, or use the planned native decoder."
                    .into(),
            ));
        }

        // PVR stores RGBA per pixel, little-endian R,G,B,A.
        let rgba = data[data_start..data_start + expected].to_vec();
        let img = RgbaImage::from_raw(width, height, rgba).ok_or_else(|| {
            AppError::InvalidFile("Failed to build RGBA image buffer.".into())
        })?;
        img.save_with_format(output, ImageFormat::Png)
            .map_err(|e| AppError::Other(format!("PNG encode failed: {e}")))?;
        Ok(expected as u64)
    }
}
